source/server/listener_manager_impl.cc

```diff
@@ -580,34 +678,74 @@ const Network::FilterChain*
+  const auto& filter_chain_local =
       source_types[envoy::api::v2::listener::FilterChainMatch_ConnectionSourceType::
                        FilterChainMatch_ConnectionSourceType_LOCAL];
 
-  auto filter_chain_external =
+  const auto& filter_chain_external =
       source_types[envoy::api::v2::listener::FilterChainMatch_ConnectionSourceType::
                        FilterChainMatch_ConnectionSourceType_EXTERNAL];
 
-  // isLocalConnection can be expensive. Call it only if LOCAL or EXTERNAL are defined.
-  const bool is_local_connection = (filter_chain_local || filter_chain_external)
-                                       ? Network::Utility::isLocalConnection(socket)
-                                       : false;
+  // isLocalConnection can be expensive. Call it only if LOCAL or EXTERNAL have entries.
+  const bool is_local_connection =
+      (!filter_chain_local.first.empty() || !filter_chain_external.first.empty())
+          ? Network::Utility::isLocalConnection(socket)
+          : false;
 
   if (is_local_connection) {
-    if (filter_chain_local) {
-      return filter_chain_local.get();
+    if (!filter_chain_local.first.empty()) {
+      return findFilterChainForSourceIpAndPort(*filter_chain_local.second, socket);
     }
   } else {
-    if (filter_chain_external) {
-      return filter_chain_external.get();
+    if (!filter_chain_external.first.empty()) {
+      return findFilterChainForSourceIpAndPort(*filter_chain_external.second, socket);
     }
   }
 
-  auto filter_chain_any =
+  const auto& filter_chain_any =
       source_types[envoy::api::v2::listener::FilterChainMatch_ConnectionSourceType::
                        FilterChainMatch_ConnectionSourceType_ANY];
 
-  return filter_chain_any.get();
+  if (!filter_chain_any.first.empty()) {
+    return findFilterChainForSourceIpAndPort(*filter_chain_any.second, socket);
+  } else {
+    return nullptr;
+  }
+}
+
+const Network::FilterChain*
+ListenerImpl::findFilterChainForSourceIpAndPort(const SourceIPsTrie& source_ips_trie,
+                                                const Network::ConnectionSocket& socket) const {
+  auto address = socket.remoteAddress();
+  if (address->type() != Network::Address::Type::Ip) {
+    address = fakeAddress();
+  }
+
+  // Match on both: exact IP and wider CIDR ranges using LcTrie.
+  const auto& data = source_ips_trie.getData(address);
+  if (data.empty()) {
+    return nullptr;
+  }
+
+  ASSERT(data.size() == 1);
+  const auto& source_ports_map = *data.back();
+  const uint32_t source_port = address->ip()->port();
+  const auto port_match = source_ports_map.find(source_port);
+
+  // Did we get a direct hit on port.
+  if (port_match != source_ports_map.end()) {
+    return port_match->second.get();
+  }
+
+  // Try port 0 if we didn't already try it (UDS).
+  if (source_port != 0) {
+    const auto any_match = source_ports_map.find(0);
+    if (any_match != source_ports_map.end()) {
+      return any_match->second.get();
+    }
+  }
```

Comment:
Did you consider making this more regular with the destination IP/port matches to support later reordering?

Reply:
What do you mean?

Reply:
Just that this merges IP/port but that isn't done in dest IP/port code, it's fine as is.